### PR TITLE
Fix issue with sparse/full config saving by custom menu logic

### PR DESCRIPTION
**Pull request by gynt**
See #204

### CODE REVIEW

File: src/components/sandbox-menu/sandbox-menu.tsx [hunk 1]
```diff
@@ -48,13 +49,26 @@ export interface SandboxArgs {
 }
 
 function saveConfig(baseUrl: string, config: Record<string, unknown>) {
+  ConsoleLogger.debug(`sandbox-menu: saveConfig: ${baseUrl}`, config);
```
Comment (TheRedDaemon) [inline src/components/sandbox-menu/sandbox-menu.tsx:52]: Intention question:  
This log will be a debug log in the browser console and not placed in the file log. This is intended? (Simple yes/no is enough)

File: src/components/sandbox-menu/sandbox-menu.tsx [hunk 2]
```diff
@@ -63,16 +77,12 @@ function saveConfig(baseUrl: string, config: Record<string, unknown>) {
...
-    .state;
-  const baselineEntries: [string, unknown][] = Object.entries(baseline)
-    .filter(([url]) => url.startsWith(baseUrl))
-    .map(([key, value]) => [key, value.modifications.value.content]);
-
-  const finalConfigEntries: Record<string, unknown> = {};
-  baselineEntries.forEach(([key, value]) => {
-    finalConfigEntries[key] = value;
-  });
+  const baseline = getStore().get(CONFIGURATION_DEFAULTS_REDUCER_ATOM);
```
Comment (TheRedDaemon) [inline src/components/sandbox-menu/sandbox-menu.tsx:80]: Ok, now I am confused. The "defaults" are the baseline?  
I thought defaults are the values provided by the options and since the "full" configuration is edited, it would aslo require the extensions states?

File: src/components/sandbox-menu/sandbox-menu.tsx [hunk 3]
```diff
@@ -88,7 +98,7 @@ function saveConfig(baseUrl: string, config: Record<string, unknown>) {
   getStore().set(CONFIGURATION_TOUCHED_REDUCER_ATOM, {
     type: 'set-multiple',
     value: Object.fromEntries(
-      Object.keys(finalConfigEntries).map((key) => [key, true]),
+      Object.keys(userConfigEntries).map((key) => [key, true]),
```
Comment (TheRedDaemon) [inline src/components/sandbox-menu/sandbox-menu.tsx:101]: So touched is only for the user values?

**Review by TheRedDaemon — CHANGES_REQUESTED** [review]

File: src/components/sandbox-menu/sandbox-menu.tsx (hunk 3 above)
Comment (gynt) [reply src/components/sandbox-menu/sandbox-menu.tsx:101]: Correct. It is a bit more difficult because it is a custom menu. In the simple case, let's say a Switch, the touched is set to true if the user changed its value (regardless of what the value is). In the Popover where you can trash your customisation for an option, this value is set back to false:
 
https://github.com/UnofficialCrusaderPatch/UCP3-GUI/blob/c6b5e70c4922c8cf04ae6c2c3986adc3c0ec58c8/src/components/ucp-tabs/config-editor/ui-elements/ui-factory/popover/ConfigPopover.tsx#L166-L169

I think in general, all the keys in touched that are true, are also in the user config. So all the things the user changed the value of in the custom menu should have `touched` set to `true`. Is this still the case with the code I wrote? Or does the custom menu set more things other than is represented in `userConfigEntries`?

File: src/components/sandbox-menu/sandbox-menu.tsx (hunk 2 above)
Comment (gynt) [reply src/components/sandbox-menu/sandbox-menu.tsx:80]: There are three sources of config:
1. Values provided by options.yml files
2. Values provided by config.yml files
3. Values provided by the user

The confusing part is that `CONFIGURATION_DEFAULTS_REDUCER_ATOM` has taken the role of representing the first two sources (see reference A). It became that way because:
1. the Customisations tab doesn't need the separation between 1 and 2 (it only displays the full config, and only needs the defaults for displaying disabling UI elements if they use `enabled: url == value`, actually maybe it doesn't need the defaults... I will make an issue) see reference B), 
2. and (reference C) the "reset this option" button uses this state to quickly go from user customisations back to what the activate extensions provide (1) and demand (2).

Maybe it is time to rename all of them :)

To clarify using set theory, configs are key value pairs.
A. Values provided by options.yml files.
B. Values provided by config.yml files.
C. Values provided by the user.

A contains all keys. B contains a subset of A (or all keys of A). C contains a subset of A (or all keys of A).
The full config is constructed by taking A, then applying the key value pairs of B, and then applying the key value pairs of C.

The full config therefore contains the same keys as A.


Code References:

A:
https://github.com/UnofficialCrusaderPatch/UCP3-GUI/blob/c6b5e70c4922c8cf04ae6c2c3986adc3c0ec58c8/src/function/extensions/state/change.tsx#L58-L74

B:
https://github.com/UnofficialCrusaderPatch/UCP3-GUI/blob/c6b5e70c4922c8cf04ae6c2c3986adc3c0ec58c8/src/components/ucp-tabs/config-editor/ui-elements/ui-factory/CreateNumberInput.tsx#L52-L56

https://github.com/UnofficialCrusaderPatch/UCP3-GUI/blob/c6b5e70c4922c8cf04ae6c2c3986adc3c0ec58c8/src/function/extensions/state/change.tsx#L64-L69


C:
https://github.com/UnofficialCrusaderPatch/UCP3-GUI/blob/c6b5e70c4922c8cf04ae6c2c3986adc3c0ec58c8/src/components/ucp-tabs/config-editor/ui-elements/ui-factory/popover/ConfigPopover.tsx#L162-L165

File: src/components/sandbox-menu/sandbox-menu.tsx [hunk 4]
```diff
@@ -48,13 +49,26 @@ export interface SandboxArgs {
 }
 
 function saveConfig(baseUrl: string, config: Record<string, unknown>) {
+  ConsoleLogger.debug(`sandbox-menu: saveConfig: ${baseUrl}`, config);
+
+  const toBeCleared = Object.entries(config)
```
Comment (gynt) [inline src/components/sandbox-menu/sandbox-menu.tsx:54]: @TheRedDaemon This is the new trashing behavior, e.g. if a key is set to undefined, then it is removed from the user config. I see I made a mistake because I didn't prepend `baseUrl` yet, as you do on line 62.

File: src/components/sandbox-menu/sandbox-menu.tsx (hunk 3 above)
Comment (TheRedDaemon) [reply src/components/sandbox-menu/sandbox-menu.tsx:101]: Nah, that is fine.

File: src/components/sandbox-menu/sandbox-menu.tsx (hunk 4 above)
Comment (TheRedDaemon) [reply src/components/sandbox-menu/sandbox-menu.tsx:54]: Ok, then please fix this.

File: src/components/sandbox-menu/sandbox-menu.tsx (hunk 2 above)
Comment (TheRedDaemon) [reply src/components/sandbox-menu/sandbox-menu.tsx:80]: Ok, then ok for this issue. Renaming could definitly help.  
And I am all for refactoring the config keeping (you already know that I would be in favour of saving every config in its own object and just providing an interface to the rest of the API which performs caches and resolves on demand, maybe even with reactive changes per value, but this would likely be a lot of work).

**Review by TheRedDaemon — CHANGES_REQUESTED** [review]

**Comment by gynt** [comment]
@TheRedDaemon In the previous code on the main branch it was possible to change the full config by settings values to undefined. This is now implemented by reconstructing the relevant parts of the full config using the baseline and the user part. This is the same thing as is done elsewhere in the code.

**Comment by gynt** [comment]
Okay what I just noticed is that when I trash Customisations, then the menu entry in the config full part of the aiSwapper becomes filled with this. It isn't present in the sparse config, and opening and closing the custom menu removes it. Do you know what is up here? Or was this already existing behavior ? :D

image

File: src/components/sandbox-menu/sandbox-menu.tsx (hunk 4 above)
Comment (TheRedDaemon) [reply src/components/sandbox-menu/sandbox-menu.tsx]: Looks good

**Comment by TheRedDaemon** [comment]
(I really start to dislike the way GIthubs reviews and comments work. Gitlab allows more freedom here... anyway)

@gynt No, this is really strange behaviour. The menu is one of these candidates that could use a "do not resolve and only available in GUI" type config value. (There would also be use for a "do not resolve but only use user config value", for example for the API replacer config).  
Anyway, since this appears after trashing, I assume it loads one of the menues of another config from a plugin **or** it is a bug, since without it being set in the user config, it would use the value of another config, or not?  
However, the strange part here is actually what is contained. When Krari orginally requested the separation of the parts in the config, I did not remove the old structure, but simply created an extra format for the "to be resolved"-config. The old config format is now saved in the menu variable and actually also loaded as the current user menu setting. (The split parts are only used for the backend or to display the other active parts of the resolved plugins). But the structure under menu in your screenshot is, as far I can tell, wrong, and the actual split config.

In short: This is either a bug that either created a faulty menu value (either as of now or when the relevant plugin was created) or someone created a plugin and then weirdly edited the raw config. Can you check the configs where these Vanilla Interpretations are stored if they have these entries under menu?

**Review by TheRedDaemon — APPROVED** [review]
